File: Difficulty: Medium/Word Search/word-search.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    bool solve(int i,int j,int ind,vector<vector<int>> &vis,vector<vector<char>>& mat, string& word){
        if(ind==word.size())
        return true;
        
        int n=mat.size(),m=mat[0].size();
        
        vis[i][j]=true;
        
        int dr[] = {-1,1,0,0};
        int dc[] = {0,0,-1,1};
    
        for(int k=0;k<4;k++){
            int row = i + dr[k];
            int col = j + dc[k];
            
            if(row>=0 && row<n && col>=0 && col<m && !vis[row][col] && mat[row][col]==word[ind] ){
                if(solve(row,col,ind+1,vis,mat,word))
                return true;
            }
        }
        
        vis[i][j]=0;
        return false;
    }
    bool isWordExist(vector<vector<char>>& mat, string& word) {
         int n=mat.size(),m=mat[0].size();
        vector<vector<int>> vis(n,vector<int>(m,0));
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(mat[i][j]==word[0])
                  if(solve(i,j,1,vis,mat,word))
                    return true;
            }
        }
        
        return false;
    }
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int n, m;
        cin >> n >> m;
        vector<vector<char>> mat(n, vector<char>(m, '*'));
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                cin >> mat[i][j];
        string word;
        cin >> word;
        Solution obj;
        bool ans = obj.isWordExist(mat, word);
        if (ans)
            cout << "true\n";
        else
            cout << "false\n";

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Word Search/word-search.java
class Solution {
    public boolean isWordExist(char[][] mat, String word) {

        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[0].length; j++) {

                if (mat[i][j] == word.charAt(0)) {
                    if (find(mat, i, j, 0, word)) {
                        return true;
                    }
                }

            }
        }

        return false;
    }

    private boolean find(char[][] mat, int row, int col, int idx, String word) {

        // If all characters are matched
        if (idx == word.length()) {
            return true;
        }

        int n = mat.length;
        int m = mat[0].length;

        // Boundary check
        if (row < 0 || col < 0 || row >= n || col >= m) {
            return false;
        }

        // Character mismatch
        if (mat[row][col] != word.charAt(idx)) {
            return false;
        }

        // Mark current cell as visited
        char temp = mat[row][col];
        mat[row][col] = '#';

        boolean result =
                find(mat, row + 1, col, idx + 1, word) ||
                find(mat, row - 1, col, idx + 1, word) ||
                find(mat, row, col + 1, idx + 1, word) ||
                find(mat, row, col - 1, idx + 1, word);

        // Restore cell
        mat[row][col] = temp;

        return result;
    }
}